src/backend/distributed/commands/database.c

```diff
@@ -242,3 +274,421 @@ PreprocessAlterDatabaseSetStmt(Node *node, const char *queryString,
+}
+
+
+#endif
+
+
+/*
+ * GenerateCreateDatabaseStatementFromPgDatabase gets the pg_database tuple and returns the
+ * CREATE DATABASE statement that can be used to create given database.
+ *
+ * Note that this doesn't deparse OID of the database and this is not a
+ * problem as we anyway don't allow specifying custom OIDs for databases
+ * when creating them.
+ */
+static char *
+GenerateCreateDatabaseStatementFromPgDatabase(Form_pg_database databaseForm)
+{
+	DatabaseCollationInfo collInfo = GetDatabaseCollation(databaseForm->oid);
+
+	StringInfoData str;
+	initStringInfo(&str);
+
+	appendStringInfo(&str, "CREATE DATABASE %s",
+					 quote_identifier(NameStr(databaseForm->datname)));
+
+	appendStringInfo(&str, " CONNECTION LIMIT %d", databaseForm->datconnlimit);
+
+	appendStringInfo(&str, " ALLOW_CONNECTIONS = %s",
+					 quote_literal_cstr(databaseForm->datallowconn ? "true" : "false"));
+
+	appendStringInfo(&str, " IS_TEMPLATE = %s",
+					 quote_literal_cstr(databaseForm->datistemplate ? "true" : "false"));
+
+	appendStringInfo(&str, " LC_COLLATE = %s",
+					 quote_literal_cstr(collInfo.datcollate));
+
+	appendStringInfo(&str, " LC_CTYPE = %s", quote_literal_cstr(collInfo.datctype));
+
+	appendStringInfo(&str, " OWNER = %s",
+					 quote_literal_cstr(GetUserNameFromId(databaseForm->datdba, false)));
+
+	appendStringInfo(&str, " TABLESPACE = %s",
+					 quote_identifier(GetTablespaceName(databaseForm->dattablespace)));
+
+	appendStringInfo(&str, " ENCODING = %s",
+					 quote_literal_cstr(pg_encoding_to_char(databaseForm->encoding)));
+
+#if PG_VERSION_NUM >= PG_VERSION_15
+	if (collInfo.datcollversion != NULL)
+	{
+		appendStringInfo(&str, " COLLATION_VERSION = %s",
+						 quote_literal_cstr(collInfo.datcollversion));
+	}
+
+	if (collInfo.daticulocale != NULL)
+	{
+		appendStringInfo(&str, " ICU_LOCALE = %s", quote_literal_cstr(
+							 collInfo.daticulocale));
+	}
+
+	appendStringInfo(&str, " LOCALE_PROVIDER = %s",
+					 quote_literal_cstr(GetLocaleProviderString(
+											databaseForm->datlocprovider)));
+#endif
+
+#if PG_VERSION_NUM >= PG_VERSION_16
+	if (collInfo.daticurules != NULL)
+	{
+		appendStringInfo(&str, " ICU_RULES = %s", quote_literal_cstr(
+							 collInfo.daticurules));
+	}
+#endif
+
+	return str.data;
+}
+
+
+/*
+ * GenerateCreateDatabaseCommandList gets a list of pg_database tuples and returns
+ * a list of CREATE DATABASE statements for all the databases.
+ *
+ * Commands in the list are wrapped by citus_internal_database_command() UDF
+ * to avoid from transaction block restrictions that apply to database commands
+ */
+List *
+GenerateCreateDatabaseCommandList(void)
+{
+	List *commands = NIL;
+
+	Relation pgDatabaseRel = table_open(DatabaseRelationId, AccessShareLock);
+	TableScanDesc scan = table_beginscan_catalog(pgDatabaseRel, 0, NULL);
+
+	HeapTuple tuple = NULL;
+	while ((tuple = heap_getnext(scan, ForwardScanDirection)) != NULL)
+	{
+		Form_pg_database databaseForm = (Form_pg_database) GETSTRUCT(tuple);
```

Comment:
Should skip the database that are not distributed (via IsOAnybjectDistributed).

Can you also add a test for that where:
* We have db1 and db2: db1 is distributed (i.e., created while the GUC is on) and db2 is not
* add_node propagates db1 but not db2.

Reply:
Fixed and added tests